cmd/kubeadm/app/cmd/join_test.go

```diff
@@ -243,3 +245,141 @@ func TestNewJoinData(t *testing.T) {
+				File:              &kubeadmapiv1beta1.FileDiscovery{},
+				TLSBootstrapToken: "",
+			},
+			validate: func(t *testing.T, opt *joinOptions) {
+				if opt.externalcfg.Discovery.TLSBootstrapToken != opt.token {
+					t.Errorf("opt.externalcfg.Discovery.TLSBootstrapToken should be set to %s", opt.token)
+				}
+				if opt.externalcfg.Discovery.File != nil {
+					t.Errorf("opt.externalcfg.Discovery.File should be set nil when len(File.KubeConfigPath) == 0")
+				}
+				if opt.externalcfg.Discovery.BootstrapToken != nil {
+					t.Errorf("opt.externalcfg.Discovery.BootstrapToken should be set nil when len(args) == 0")
+				}
+				if opt.externalcfg.ControlPlane != nil {
+					t.Errorf("opt.externalcfg.ControlPlane should be set nil when !opt.controlPlane")
+				}
+			},
+		},
+		{
+			description:  "len(opt.token) > 0 and len(args) > 0 and controlPlane is true",
+			args:         []string{"1.2.3.4:6443"},
+			token:        "abcdef.0123456789abcdef",
+			controlPlane: true,
+			discovery: kubeadmapiv1beta1.Discovery{
+				BootstrapToken:    &kubeadmapiv1beta1.BootstrapTokenDiscovery{},
+				File:              &kubeadmapiv1beta1.FileDiscovery{KubeConfigPath: "/tmp/kubeconfig"},
+				TLSBootstrapToken: "abcdef.0123456789",
+			},
+			validate: func(t *testing.T, opt *joinOptions) {
+				if opt.externalcfg.Discovery.TLSBootstrapToken != "abcdef.0123456789" {
+					t.Errorf("opt.externalcfg.Discovery.TLSBootstrapToken shouldn't be set to %s", opt.externalcfg.Discovery.TLSBootstrapToken)
+				}
+				if opt.externalcfg.Discovery.BootstrapToken.Token != opt.token {
+					t.Errorf("opt.externalcfg.Discovery.BootstrapToken should be set nil when len(args) == 0")
+				}
+				if opt.externalcfg.Discovery.File == nil {
+					t.Errorf("opt.externalcfg.Discovery.File shouldn't be set nil when len(File.KubeConfigPath) != 0")
+				}
+				if opt.externalcfg.Discovery.BootstrapToken.APIServerEndpoint != "1.2.3.4:6443" {
+					t.Errorf("opt.externalcfg.Discovery.BootstrapToken.APIServerEndpoint should be set args[0]")
+				}
+				if opt.externalcfg.ControlPlane == nil {
+					t.Errorf("opt.externalcfg.ControlPlane shouldn't be set nil when opt.controlPlane is true")
+				}
+			},
+		},
+	}
+	for _, rt := range tests {
+		t.Run(rt.description, func(t *testing.T) {
+			joinOptions := initJoinOptions(rt.token, rt.discovery, rt.controlPlane)
+
+			amendJoinOptions(joinOptions, rt.args)
+			if rt.validate != nil {
+				rt.validate(t, joinOptions)
+			}
+		})
+	}
+}
+
+func TestLoadJoinConfiguration(t *testing.T) {
+	var tests = []struct {
+		description string
+		opt         *joinOptions
+		validate    func(*testing.T, *kubeadmapi.JoinConfiguration)
+		expectedErr bool
+	}{
+		{
+			description: "configutil.LoadOrDefaultJoinConfiguration failed should return error",
+			opt: &joinOptions{
+				cfgPath: "/tmp/cfgPath",
+			},
+			expectedErr: true,
+		},
+		{
+			description: "set cfg.NodeRegistration name and CRISocket",
```

Comment:
This case is not very interesting, since we don't load the config from a file and then overriding the `Name` and `CRISocket` fields only. We simply use `externalcfg` as is which is a different and also valid test case, that needs to be in here too.